### PR TITLE
[SU-215] Version related set tables with tables

### PR DESCRIPTION
**Pull request by nawatts**
Currently, versions of individual data tables can be saved and restored. This changes set tables so that they may be versioned with the table that they reference. This makes it easy to have the restored set table reference the restored referenced table.

To enable data table versioning, go to the feature preview page (`/#feature-preview`) and check the box for data table versioning.

https://user-images.githubusercontent.com/1156625/193836501-da9305a9-23a3-4508-97d7-6d75e8b4481b.mov

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -68,7 +68,7 @@
     "preinstall": "node .hooks/check-engine-light.js",
     "setenv": "echo REACT_APP_VERSION=$(git rev-parse HEAD) REACT_APP_BUILD_TIMESTAMP=$(date -u \"+%s000\")",
     "start": "env $(yarn setenv) react-app-rewired start",
-    "test": "react-app-rewired test",
+    "test": "TZ=UTC react-app-rewired test",
```
Comment (nawatts) [inline package.json:71]: With this, tests use the same time zone locally as in CircleCI.

File: src/libs/ajax.js [hunk 2]
```diff
@@ -890,6 +890,8 @@ const Workspaces = signal => ({
         return fetchRawls(`${root}/entities/delete`, _.mergeAll([authOpts(), jsonBody(entities), { signal, method: 'POST' }]))
       },
 
+      getEntities: entityType => fetchRawls(`${root}/entities/${entityType}`, _.merge(authOpts(), { signal })).then(r => r.json()),
```
Comment (MatthewBemis) [inline src/libs/ajax.js:892]: This is the Rawls endpoint that is (or was) susceptible to OOM issues for very large data tables. I believe AJ added in some limits to protect against that case by limiting how many rows would be returned, but just a heads up that this may introduce some unexpected behavior for very large tables. Since it's a POC I don't think it's worth handling at this time.

**Review by MatthewBemis — APPROVED** [review]

File: src/libs/ajax.js (hunk 2 above)
Comment (nawatts) [reply src/libs/ajax.js:892]: I haven't tried it with large tables. That's unfortunate if that endpoint doesn't work as advertised. Though since creating a version fetches entity metadata, this could easily be changed to use [entityQuery](https://rawls.dsde-prod.broadinstitute.org/#/entities/entityQuery).

Comment (MatthewBemis) [reply src/libs/ajax.js:892]: That's good that it can be switched over to entityQuery if necessary- I do think AJ wants to eventually remove the listEntitiesOfType endpoint entirely. The limit they added is pretty high, but I think some of the users we're targeting for the feature preview may have tables large enough to hit it.

Here's the relevant Rawls PR that added the limit: https://github.com/broadinstitute/rawls/pull/1647

Comment (nawatts) [reply src/libs/ajax.js:892]: Switched endpoints in 4a16b32f6478d69ff5bb2fd2fdb6617f81c4b3c4.

Comment (MatthewBemis) [reply src/libs/ajax.js:892]: Nice! That commit LGTM, still a 👍

**Comment by cahrens** [comment]
Some usability things:

1) I was confused because when I selected "Restore", the table did not go back to the version I was restoring to. Instead, I guess it create a new table? I feel like "Restore" is the wrong word for that action.

2) Why are the buttons laid out on top of each other? To me it looks odd, and I was also confused because I expected to see the table when I clicked on the tab, instead of having to download it and look at it outside of Terra.

image

File: src/components/data/data-table-versions.test.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  it('renders restore button and confirms restore', () => {
+    const onRestore = jest.fn()
+    const { getByTestId, getByText } = render(h(DataTableVersion, { version: testVersion, onDelete: jest.fn(), onRestore }))
+
+    const restoreButton = getByText('Restore')
+    fireEvent.click(restoreButton)
+
+    expect(getByText(/This version will be restored to a new data table/)).toBeTruthy()
+
+    const confirmRestoreButton = getByTestId('confirm-restore')
```
Comment (cahrens) [inline src/components/data/data-table-versions.test.js]: Why don't you just get the button by the Restore button text? Is it because there are 2 buttons with that text on the page?

**Comment by cahrens** [comment]
@nawatts can you share with me (cahrens@gmail.com) a workspace that has a set table? I don't understand how to create that structure (I don't know much about how our data tab works).

Speaking of this, could you also make sure that the team that owns this tab (Analysis Journeys?) is up-to-date on these changes?

File: src/libs/data-table-versions.js [hunk 4]
```diff
@@ -13,32 +14,102 @@ import * as Utils from 'src/libs/utils'
 
 export const dataTableVersionsPathRoot = '.data-table-versions'
 
-export const saveDataTableVersion = async (workspace, entityType, { description = null } = {}) => {
+const getAllEntities = async (workspace, entityType) => {
+  const { workspace: { namespace, name } } = workspace
+
+  const pageSize = 100_000
```
Comment (cahrens) [inline src/libs/data-table-versions.js:19]: I don't recognize this syntax (100_000). Can you explain?

Comment (nawatts) [reply src/libs/data-table-versions.js:19]: Underscores can be used as separators in numeric literals. I think it's easier to read when there's a lot of zeros.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#numeric_separators

File: src/components/data/data-table-versions.test.js (hunk 3 above)
Comment (nawatts) [reply src/components/data/data-table-versions.test.js]: > Is it because there are 2 buttons with that text on the page?

Yes. Both the button on the page and the button in the confirmation button are labeled "Restore".

**Comment by nawatts** [comment]
> I was confused because when I selected "Restore", the table did not go back to the version I was restoring to. Instead, I guess it create a new table? I feel like "Restore" is the wrong word for that action.

Yes. The "Restore" confirmation modal shows the new table name. We decided against restoring to the original table to avoid overwriting data and to avoid having to handle broken references (though that's probably not as much of an issue now that we're versioning set tables with the referenced tables).

We could probably come up with a different label for this function. "Import"?

> Why are the buttons laid out on top of each other? To me it looks odd, and I was also confused because I expected to see the table when I clicked on the tab, instead of having to download it and look at it outside of Terra.

There was no thought put into that layout. Do you have a suggestion for an alternative? Maybe a horizontal row with "Download" and "Restore" on the left and "Delete" on the right?

I don't think there's a good way to view the version's data in Terra without storing versions as data tables, in which case they would be editable.

> can you share with me a workspace that has a set table? I don't understand how to create that structure (I don't know much about how our data tab works).

Yes, I'll share one. You can create a set by selecting data and clicking "Save selection as a set" in the Edit menu. Sets are also automatically created when running a workflow on some data.

> Speaking of this, could you also make sure that the team that owns this tab (Analysis Journeys?) is up-to-date on these changes?

I demoed this to the Analysis Journeys team yesterday.

**Comment by nawatts** [comment]
Shared `general-dev-billing-account/set-tables`. Also found a bug with sets of sets. Looks like the TSV upload doesn't support membership TSVs for nested sets.

**Comment by cahrens** [comment]
I agree that with the current functionality, "Import" would make more sense than "Restore". And yes, I think the buttons should be laid out horizontally instead of vertically. The vertical layout makes me think something is supposed to be beside them.

I think a read-only view of data tables should be introduced because one would really want to view the data in Terra before choosing to import it. But that's obviously a different story.

**Comment by cahrens** [comment]
Could use some more space between the checkbox and the label. Though I suppose that's a pre-existing issue with the shared component?

image

**Comment by nawatts** [comment]
I'll change the button text and layout.

Since uploading membership TSVs is broken for nested sets, looks like I'll have to switch to storing JSON in the version files. That would make it easier to add a preview in the future.

**Comment by cahrens** [comment]
@nawatts I'm getting an error when trying to import a saved version (even a new one that I just created).

image

**Comment by nawatts** [comment]
Ah, looks like #3403 changed the behavior of the `getObjectPreview` method (and broke file previews) so when I rebased this branch it stopped working.

https://github.com/DataBiosphere/terra-ui/pull/3403/files#diff-7a78ee750caec6a6aa1725bf8bdb8a59144ebf6e9c92f26fd8b3903be5b71b02R28

**Comment by nawatts** [comment]
#3443 contains a fix for the `getObjectPreview` function.

**Comment by nawatts** [comment]
Rebased after #3443 was merged, so that error should be resolved now.

File: src/components/data/data-table-versions.test.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      const { getByText } = renderResult
+      expect(getByText('A version of samples')).toBeTruthy()
+    })
+  })
+
+  it('renders restore button and confirms restore', () => {
+    const onRestore = jest.fn()
+    const { getByTestId, getByText } = render(h(DataTableVersion, { version: testVersion, onDelete: jest.fn(), onRestore }))
+
+    const restoreButton = getByText('Import')
```
Comment (cahrens) [inline src/components/data/data-table-versions.test.js]: Nit: it might be nice to change "restore" to "import" in the code as well so it isn't left with this legacy of having started as "restore".

File: src/libs/data-table-versions.js [hunk 6]
```diff
@@ -89,25 +123,44 @@ export const restoreDataTableVersion = async (workspace, version) => {
...
+    const restoredSetTableName = _.replace(version.entityType, restoredTableName, setTableName)
+    const restoredSetTableMemberType = restoredSetTableName.slice(0, -4)
+
+    const setTableEntities = JSON.parse(await zip.file(`json/${setTableName}.json`).async('text'))
+    const setTableEntityUpdates = _.map(({ name, attributes }) => ({
+      entityType: restoredSetTableName,
+      name,
+      operations: Object.entries({
+        ...attributes,
+        [`${originalSetTableMemberType}s`]: _.update(
```
Comment (cahrens) [inline src/libs/data-table-versions.js:152]: I'm confused how this works with the "s" stuck on at the end-- can you briefly explain?

File: src/pages/workspaces/workspace/Data.js [hunk 7]
```diff
@@ -688,13 +692,23 @@ const WorkspaceData = _.flow(
...
+                          loadDataTableVersions(type.endsWith('_set') ? _.replace(/(_set)+$/, '', type) : type)
                         }
                       })
                     })
                   }),
                   isShowingVersionHistory && h(DataTableVersions, {
-                    ...dataTableVersions[type],
+                    ...Utils.cond(
+                      [type.endsWith('_set'), () => {
+                        const referencedType = _.replace(/(_set)+$/, '', type)
```
Comment (cahrens) [inline src/pages/workspaces/workspace/Data.js]: `_.replace(/(_set)+$/, '', type)` seems to be in a few different places. Is it work making it a utility method?

File: src/libs/data-table-versions.js [hunk 8]
```diff
@@ -1,44 +1,77 @@
...
+  const { workspace: { namespace, name } } = workspace
+
+  const pageSize = 100_000
+  const firstPageResponse = await Ajax().Workspaces.workspace(namespace, name).paginatedEntitiesOfType(entityType, { page: 1, pageSize })
+
+  let entities = firstPageResponse.results
+  const numPages = firstPageResponse.resultMetadata.filteredPageCount
+
+  for (let page = 2; page <= numPages; page += 1) {
+    const pageResponse = await Ajax().Workspaces.workspace(namespace, name).paginatedEntitiesOfType(entityType, { page, pageSize })
```
Comment (cahrens) [inline src/libs/data-table-versions.js:26]: It looks like this pagination code isn't covered by unit tests. Could you add a test case for it?

image

**Comment by cahrens** [comment]
@nawatts Just a couple more comments/requests… otherwise it LGTM.

File: src/libs/data-table-versions.js (hunk 6 above)
Comment (nawatts) [reply src/libs/data-table-versions.js:152]: This is how Terra names fields in set tables. If there's a "sample" table, the set table is named "sample_set" and contains a column "samples" with a list of references to "sample".

File: src/components/data/data-table-versions.test.js (hunk 5 above)
Comment (nawatts) [reply src/components/data/data-table-versions.test.js]: Done in https://github.com/DataBiosphere/terra-ui/pull/3434/commits/5bd56538295ad92423852dd871b56458ac030230.

File: src/pages/workspaces/workspace/Data.js (hunk 7 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/Data.js]: Done in [f2838c5](https://github.com/DataBiosphere/terra-ui/pull/3434/commits/f2838c53ada51d537fa28ec99aa5a5068d03535d).

File: src/libs/data-table-versions.js (hunk 8 above)
Comment (nawatts) [reply src/libs/data-table-versions.js:26]: Done in https://github.com/DataBiosphere/terra-ui/pull/3434/commits/499f342644579d87f6bd0d11a9d76878880c9b2b.

File: src/components/data/data-utils.js [hunk 9]
```diff
@@ -64,6 +64,8 @@ export const getUserProjectForWorkspace = async workspace => (workspace && await
 
 const isUri = datum => _.startsWith('gs://', datum) || _.startsWith('dos://', datum) || _.startsWith('drs://', datum)
 
+export const getRootTypeForSetTable = tableName => _.replace(/(_set)+$/, '', tableName)
```
Comment (cahrens) [inline src/components/data/data-utils.js:67]: @MatthewBemis FYI, I think your code also had this regexp.

**Review by cahrens — APPROVED** [review]
Thanks @nawatts!
